pkg/controllers/allocation/scheduling/scheduler.go

```diff
@@ -147,3 +156,56 @@ func IsSchedulable(pod *v1.Pod, node *v1.Node) bool {
 	// TODO, support node affinity
 	return true
 }
+
+type schedulerMetricsDecorator struct {
+	scheduler                Scheduler
+	scheduleTimeHistogramVec *prometheus.HistogramVec
+}
+
+func DecorateSchedulerMetrics(scheduler Scheduler) Scheduler {
+	scheduleTimeHistogramVec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metrics.KarpenterNamespace,
+			Subsystem: "allocation_controller",
+			Name:      "scheduling_duration_seconds",
+			Help:      "Duration of scheduling process in seconds. Broken down by provisioner and result.",
+			Buckets:   metrics.DurationBuckets(),
+		},
+		[]string{metrics.ProvisionerLabel, metrics.ResultLabel},
+	)
+	crmetrics.Registry.MustRegister(scheduleTimeHistogramVec)
+
+	return &schedulerMetricsDecorator{scheduler: scheduler, scheduleTimeHistogramVec: scheduleTimeHistogramVec}
+}
+
+func (s *schedulerMetricsDecorator) Solve(ctx context.Context, provisioner *v1alpha3.Provisioner, pods []*v1.Pod) ([]*Schedule, error) {
+	startTime := time.Now()
+	schedules, scheduleErr := s.scheduler.Solve(ctx, provisioner, pods)
+	durationSeconds := time.Since(startTime).Seconds()
+
+	result := "success"
```

Comment:
I'm noticing that controller runtime separate success and error cases into separate metrics https://github.com/kubernetes-sigs/controller-runtime/blob/v0.10.0/pkg/internal/controller/metrics/metrics.go#L37. Thoughts on following the same pattern?

Reply:
I think using labels better encapsulates the metric information.  An analogy in code would be passing around a `struct` vs passing around multiple related but individual variables.

Reply:
Do you have some examples from similar kubernetes projects?

Reply:
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.10.0/pkg/internal/controller/metrics/metrics.go#L30-L33 brakes out the metric by result classification.

Reply:
Great thanks!